compiler/rustc_driver_impl/src/lib.rs

```diff
@@ -439,6 +439,14 @@ fn run_compiler(
                 return early_exit();
             }
 
+            rustc_interface::export::write_interface(tcx);
+
+            if sess.opts.output_types.contains_key(&OutputType::Interface)
+                && sess.opts.output_types.len() == 1
+            {
+                return early_exit();
+            }
```

Comment:
I think the interface shouldn't be written until after `tcx.ensure().analysis(())`. We shouldn't emit an interface file with type errors.

Reply:
Update: HIR pretty printer could no longer be used as an interface generator, since the new implementation of attributes uses pseudo-code. Therefore, I have moved the interface generator to the AST. This has led to the fact that non-exportable items cannot be filtered out, but I don't think this is a major issue at the moment.